File: BOJ/DynamicProgramming/Kotlin/2921.kt
import java.util.*

fun main(args: Array<String>) = with(Scanner(System.`in`)) {
	var dp  = Array(1001) { Array(1001) { 0L } }
	var dp2 = Array(1001) { Array(1001) { 0L } }

	dp[0][0] = 1L
	for (i in 0L..1000L)
		dp2[i.toInt()][1] = i

	for (H in 1..1000) {
		var current = 0L
		for (W in 0..1000) {
			current += dp[W][H-1]
			dp[W][H] = current
		}
	}

	for (H in 1..1000) {
		var current = 0L
		for (W in 0..1000) {
			current += dp2[W][H-1] + W*dp[W][H]
			dp2[W][H] = current
		}
	}

	for (H in 1..1000) {
		for (W in 0..1000) {
			print("(")
			print(dp[W][H])
			print(",")
			print(dp2[W][H])
			print(")")
			print(" ")
		}
		println("")
	}
	val N = nextInt()
	print(dp2[N][N])
}

File: BOJ/DynamicProgramming/Python/2921.py
dp  = [ [0] * 1002 for i in range(1002) ]
dp2 = [ [0] * 1002 for i in range(1002) ]

acc  = [ [0] * 1002 for i in range(1002) ]
acc2 = [ [0] * 1002 for i in range(1002) ]

dp[0][0] = 1
for i in range(1001):
    dp2[i][1] = i

for H in range(1, 1001):
    current = 0
    for W in range(1001):
        current += dp[W][H-1]
        dp[W][H] = current
    print

for H in range(2, 1001):
    current = 0
    for W in range(1001):
        current += dp2[W][H-1] + W*dp[W][H]
        dp2[W][H] = current
        

n=int(input())
print(dp2[n][n] + 1)
